File: app/src/main/java/linusfessler/alarmtiles/alarm/AlarmFragment.java
package linusfessler.alarmtiles.alarm;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import javax.inject.Inject;

import io.reactivex.disposables.CompositeDisposable;
import linusfessler.alarmtiles.core.App;
import linusfessler.alarmtiles.databinding.FragmentAlarmBinding;

public class AlarmFragment extends Fragment {

    @Inject
    AlarmViewModelFactory viewModelFactory;

    private AlarmViewModel viewModel;
    private FragmentAlarmBinding binding;
    private final CompositeDisposable disposable = new CompositeDisposable();

    @Override
    public void onCreate(@Nullable final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        ((App) requireActivity().getApplicationContext())
                .getAlarmComponent()
                .inject(this);

        viewModel = new ViewModelProvider(this, viewModelFactory).get(AlarmViewModel.class);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull final LayoutInflater inflater, @Nullable final ViewGroup container, @Nullable final Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);

        binding = FragmentAlarmBinding.inflate(inflater, container, false);

        viewModel.getAlarm().observe(getViewLifecycleOwner(), alarm -> {
            if (alarm == null) {
                return;
            }
            binding.alarm.setEnabled(alarm.isEnabled());
            binding.alarm.setOnClickListener(view -> viewModel.toggle(alarm));
        });

        viewModel.getTileLabel().observe(getViewLifecycleOwner(), binding.alarm::setLabel);

        return binding.getRoot();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        disposable.clear();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        disposable.dispose();
    }
}

File: app/src/main/java/linusfessler/alarmtiles/alarm/AlarmFragment.kt
package linusfessler.alarmtiles.alarm

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.Fragment
import io.reactivex.disposables.CompositeDisposable
import linusfessler.alarmtiles.R
import linusfessler.alarmtiles.core.App
import linusfessler.alarmtiles.databinding.FragmentAlarmBinding
import javax.inject.Inject

class AlarmFragment : Fragment() {
    @Inject
    lateinit var viewModel: AlarmViewModel

    private lateinit var binding: FragmentAlarmBinding
    //    private lateinit var startDialog: AlarmStartDialog
    private lateinit var descriptionDialog: AlertDialog

    private val disposable = CompositeDisposable()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        (requireActivity().applicationContext as App)
                .alarmComponent
                .inject(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        binding = FragmentAlarmBinding.inflate(inflater, container, false)
//        startDialog = AlarmStartDialog(requireContext(), inputMethodManager, viewModel)
        descriptionDialog = AlertDialog.Builder(requireContext())
                .setTitle(R.string.alarm)
                .setMessage(R.string.alarm_description)
                .setCancelable(true)
                .create()

        binding.alarm.setOnClickListener {
            disposable.add(viewModel.alarm
                    .firstElement()
                    .subscribe {
                        if (it.isEnabled) {
                            viewModel.dispatch(AlarmEvent.Disable())
                        } else {
                            viewModel.dispatch(AlarmEvent.Enable(8, 0))
//                            startDialog.show()
                        }
                    })
        }

        binding.alarm.setOnLongClickListener {
            descriptionDialog.show()
            true
        }

        disposable.add(viewModel.alarm
                .subscribe {
                    if (it.isEnabled) {
                        // It's possible that the alarm was set through the quick settings while the start dialog is shown, dismiss it in this case
//                        startDialog.dismiss()
                    }
                    binding.alarm.isEnabled = it.isEnabled
                })

        disposable.add(viewModel.timeLeft
                .subscribe {
                    binding.alarm.setSubtitle(it)
                })

        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        disposable.clear()
        binding.alarm.setOnClickListener(null)
        binding.alarm.setOnLongClickListener(null)
    }

    override fun onDestroy() {
        super.onDestroy()
        disposable.dispose()
//        startDialog.dismiss()
        descriptionDialog.dismiss()
    }
}